packages/nimble-components/src/select/index.ts

```diff
@@ -969,39 +1017,131 @@ export class Select
+            ) {
+                return previousElement;
+            }
+            previousElement = previousElement.previousElementSibling;
+        }
+        return null;
+    }
+
+    private getNextVisibleOptionOrGroup(
+        element: HTMLElement
+    ): ListOption | ListOptionGroup | null {
+        let nextElement = element.nextElementSibling;
+        while (nextElement) {
+            if (
+                (nextElement instanceof ListOption
+                    || nextElement instanceof ListOptionGroup)
+                && isOptionOrGroupVisible(nextElement)
+            ) {
+                return nextElement;
+            }
+            nextElement = nextElement.nextElementSibling;
+        }
+        return null;
+    }
+
+    private isOptionHiddenOrFilteredOut(option: ListOption): boolean {
+        if (option.hidden) {
+            return true;
+        }
+        return !this.filterMatchesText(option.text);
+    }
+
+    private filterMatchesText(text: string): boolean {
+        const filter = this.filter.toLowerCase();
+        const normalizedFilter = diacriticInsensitiveStringNormalizer(filter);
+        return diacriticInsensitiveStringNormalizer(text).includes(
+            normalizedFilter
+        );
+    }
+
     /**
      * Filter available options by text value.
      *
      * @public
      */
     private filterOptions(): void {
-        const filter = this.filter.toLowerCase();
+        if (!this.$fastController.isConnected) {
+            return;
+        }
 
-        if (filter) {
-            this.filteredOptions = this.options.filter(option => {
-                const normalizedFilter = diacriticInsensitiveStringNormalizer(filter);
-                return (
-                    !option.hidden
-                    && diacriticInsensitiveStringNormalizer(option.text).includes(
-                        normalizedFilter
-                    )
+        const filteredOptions: ListOption[] = [];
+        for (const element of this.slottedOptions) {
+            if (element instanceof ListOptionGroup) {
+                if (element.hidden) {
+                    break; // no need to process hidden groups
+                }
+                const groupOptions = this.getGroupOptions(element);
+                const groupMatchesFilter = this.filterMatchesText(
+                    element.labelContent
                 );
-            });
-        } else {
-            this.filteredOptions = this.options.filter(
-                option => !option.hidden
-            );
+                groupOptions.forEach(option => {
+                    option.visuallyHidden = groupMatchesFilter
+                        ? false
+                        : this.isOptionHiddenOrFilteredOut(option);
+                    if (!option.visuallyHidden) {
+                        filteredOptions.push(option);
+                    }
+                });
+            } else if (isNimbleListOption(element)) {
```

Comment:
Is there a strategy for when to use `isNimbleListOption(element)` vs `instanceof ListOption`? Same for `instanceof ListOptionGroup`. I think the code would be easier to read if we stayed at the same level of abstraction within a given file / method. I would vote for helper functions everywhere, and only doing instanceof checks within the helper function.

Reply:
Agreed and updated as suggested.